src/shared/sleep-config.c

```diff
@@ -141,6 +144,134 @@ int read_battery_capacity_percentage(void) {
+        if (battery_discharge_rate <= 0 || battery_discharge_rate >= 200)
+               return 0;
+               /* battery discharge rate is not in a valid range */
+        log_debug("Battery discharge rate is in valid range");
+        return 1;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *hash_id_discharge_rate = NULL;
+        const char *p, *filepath;
+        uint8_t current_hash_id[SHA256_DIGEST_SIZE];
+        char *stored_hash_id = NULL, *stored_discharge_rate = NULL, *h = NULL;
+        int discharge_rate, r;
+
+        filepath = "/var/lib/systemd/sleep/battery_discharge_percentage_rate_per_hour";
+
+        r = read_one_line_file(filepath, &hash_id_discharge_rate);
+        log_debug("Stored hash_id and discharge_rate is %s", hash_id_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read discharge rate from %s: %m", filepath);
+
+        r = get_system_battery_identifier_hash(current_hash_id);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+
+        p = hash_id_discharge_rate;
+        r = extract_many_words(&p, " ", 0, &stored_hash_id, &stored_discharge_rate, NULL);
+        log_debug("stored hash_id is %s and stored discharge_rate is %s", stored_hash_id, stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse hash_id and discharge_rate read from %s location: %m", filepath);
+
+        h = hexmem(current_hash_id, sizeof(current_hash_id));
+        if (!strcmp_ptr(h, stored_hash_id))
+               return log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "Current dentifier do not match stored identifier: %m");
+        log_debug("hash id match");
+
+        r = safe_atoi(stored_discharge_rate, &discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse discharge rate read from %s location: %m", filepath);
+
+        if (!battery_discharge_rate_in_range(discharge_rate))
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery discharge percentage rate per hour: %m");
+
+        return discharge_rate;
+}
+
+int put_battery_discharge_rate(int estimated_battery_discharge_rate) {
+        uint8_t system_hash_id[SHA256_DIGEST_SIZE];
+        const char *filepath;
+        char *p = NULL, *h = NULL;
+        int r = 0;
+
+        if (!battery_discharge_rate_in_range(estimated_battery_discharge_rate))
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery discharge percentage rate per hour: %m");
+
+        r = get_system_battery_identifier_hash(system_hash_id);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+
+        h = hexmem(system_hash_id, sizeof(system_hash_id));
+        log_debug("system_hash_id is %s", h);
+        p = strjoin(h, " ", estimated_battery_discharge_rate);
```

Comment:
no need to strjoin here, just pass them individually to _filef below with `"%s %s"`

Reply:
Yeah right. Missed that. Fixed